CryptKeeperZK/crypt-keeper-extension
# feat: request verifiable presentation from user (#509)

## Explanation

This change implements Verifiable Presentations, a way for users to present Verifiable Credentials to verifiers in an authentic way. The lifecycle is as follows:
1. Web application requests Verifiable Credentials from a user, with a certain request message detailing what type of credentials the service is looking for.
2. A popup allows users to select Verifiable Credentials they wish to share with the application.
3. Cryptkeeper generates a Verifiable Presentation from the selected credentials.
4. Users may either send the Verifiable Presentation without a signature or use their Metamask Wallet to sign the presentation with an Ethereum keypair.
5. The web application verifies that the Verifiable Presentation is authentic and that it contains the credentials it is looking for (not implemented in Cryptkeeper). 

- [x] Create popup allowing users to select Verifiable Credentials
- [x] Generate Verifiable Presentation from selected credentials
- [x] Allow users to sign the Verifiable Presentation
- [x] Connect Verifiable Presentation request and submission to event lifecycle 
- [x] Integrate Verifiable Presentations into demo 
- [x] Add tests for components and hooks
- [x] Make Verifiable Credential Items in popup similar to IdentityList Items

## More Information
See [this change](https://github.com/CryptKeeperZK/crypt-keeper-extension/pull/674) for reference on integration with displaying Verifiable Credentials.

## Screenshots/Screencaps
View a request for Verifiable Credentials and select ones you want to share:
<img width="375" alt="Screen Shot 2023-09-13 at 4 14 33 PM" src="https://github.com/CryptKeeperZK/crypt-keeper-extension/assets/22803219/08cda9d9-056c-41d6-b521-5f7c7354b53f">

Choose to either sign a Verifiable Presentation or send it in the clear:
<img width="381" alt="Screen Shot 2023-09-13 at 4 15 48 PM" src="https://github.com/CryptKeeperZK/crypt-keeper-extension/assets/22803219/9345e803-0c5b-45ae-9ea9-49cf8f15507f">

View a Verifiable Presentation request in full page window:
<img width="1116" alt="Screen Shot 2023-09-13 at 4 16 08 PM" src="https://github.com/CryptKeeperZK/crypt-keeper-extension/assets/22803219/70192f47-52ff-4359-b8d6-119952e20a40">

Sign your Verifiable Presentation with metamask:
<img width="360" alt="Screen Shot 2023-09-11 at 12 11 24 PM" src="https://github.com/CryptKeeperZK/crypt-keeper-extension/assets/22803219/367176ae-e393-43fd-992a-4c47e4830fe5">


## Manual Testing Steps
1. Load the demo
2. Add some Verifiable Credentials to your wallet
3. Click on the button to "Generate a Verifiable Presentation
4. Select the credentials you would like to share
5. Connect your Metamask
6. Authorize the signature
7. Check that the demo has received your signed Verifiable Presentation

## Pre-Merge Checklist


> PR template source from [github.com/MetaMask](https://github.com/MetaMask)
